RawHTML component: Allow multiple children

andrewserong opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->

<!-- Gutenberg's license is in the process of updating to be dual-licensed under the GPL and MPL. As part of that transition, all new contributions are dual-licensed. For more information, see: https://github.com/WordPress/gutenberg/blob/trunk/LICENSE.md -->

## Description
<!-- Please describe what you have changed or added -->
This PR updates the RawHTML component to support passing multiple children instead of treating `children` as a string.

This means that if you wish to use the RawHTML component and include raw HTML from a few different variables, you don't need to manually concatenate those string before passing them to the component.

I have also updated the tests for the RawHTML component so that it uses React Testing Library and tests the output of the component, instead of the internal state of the React element.

## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->

Running tests:

```
npm run test-unit -- --testPathPattern packages/element/src/test/raw-html.js
```

For a concrete example of the issue that I ran into that led to this PR, I was playing around with adding an extra string to the `RawHTML` component used in the post content block's ReadOnlyContent component (rendered when a Post Content block appears as a child of the Query loop block). [Around this line](https://github.com/wordpress/gutenberg/blob/d079ba2c16495d6d976f5e08a7b5d259c3f13074/packages/block-library/src/post-content/edit.js#L36).

Example code:

```
		<div { ...blockProps }>
			<RawHTML key="html">
				{ content?.rendered } { '<p>Another paragraph</p>' }
			</RawHTML>
		</div>
```

The above code, before this PR, results in stray `,` characters being rendered, as the `children` array is joined into a string using `,` characters as a separator. With this PR applied, we concatenate the strings before passing to `dangerouslySetInnerHTML` so there are no stray comma characters.

If you wish to reproduce this, use the above code in line 36 of `/packages/block-library/src/post-content/edit.js` and using a block-based theme like TT1-Blocks, publish a post. Then, open up the site editor and view the post content block within a Query loop block. You should be able to then see the differences in the before/after screenshots below.

## Screenshots

Note the stray comma characters in the Before screenshot, and that they no longer appear in the After screenshot.

| Before | After |
| --- | --- |
| ![image](https://user-images.githubusercontent.com/14988353/136877080-a72474ba-9050-4ef9-9c71-bf3d6bf72067.png) | ![image](https://user-images.githubusercontent.com/14988353/136877093-189189b3-69cb-44de-9bbc-b94ba2429208.png) |

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->
Bug fix (non-breaking change which fixes an issue)

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

andrewserong committed:
RawHTML component: Allow multiple children

andrewserong commented:
CC @ciampo, just an FYI as this is a component used in various places. It isn't part of the Components package (so I haven't added you as a reviewer), but just thought I'd keep you in the loop 🙂

ciampo commented on packages/element/src/raw-html.js:
```diff
@@ -12,16 +12,20 @@ import { createElement } from './react';
+	const rawHtml = Array.isArray( children ) ? children.join( '' ) : children;
+
 	// The DIV wrapper will be stripped by serializer, unless there are
 	// non-children props present.
```
```js
// The DIV wrapper will be stripped by serializer, unless there are
// non-children props present.
```

Is this still holding true? From the unit tests, it seems like the wrapper `div` is always added to the output

andrewserong replied:
Good catch, it looks like the `div` has always been added to the output, so I've removed the comment.

ciampo commented on packages/element/src/test/raw-html.js:
```diff
@@ -1,7 +1,7 @@
  * External dependencies
  */
-import { shallow } from 'enzyme';
+import { render } from '@testing-library/react';
```
Thank you for switching to RTL !

ciampo commented on packages/element/src/raw-html.js:
```diff
@@ -12,16 +12,20 @@ import { createElement } from './react';
  *
- * @param {RawHTMLProps} props Children should be a string of HTML. Other props
- *                             will be passed through to div wrapper.
+ * @param {RawHTMLProps} props Children should be a string of HTML or an array
```
Is there ever a scenario where something different than strings is passed as children of `RawHTML`?

ciampo replied:
Let's continue this conversation in https://github.com/WordPress/gutenberg/pull/35532#discussion_r726973260

ciampo reviewed (commented):
Thank you for the ping! Always interesting :) 

I don't really have any context for this component, and so I left a few comments which hopefully aren't silly.

Also, what's the need for such a component? We should make sure that it's fed raw HTML only from trusted sources (e.g. not plugins / users / unknown API responses...)

gziolo commented on packages/element/src/raw-html.js:
```diff
@@ -12,16 +12,20 @@ import { createElement } from './react';
  */
 export default function RawHTML( { children, ...props } ) {
+	// Concatenate into a single string if children is an array.
+	const rawHtml = Array.isArray( children ) ? children.join( '' ) : children;
```
It should go through `React.Children` utils first as explained in https://reactjs.org/docs/react-api.html#reactchildren. It's a more complex structure, so better to account for other cases if this API needs to be more flexible.


https://reactjs.org/docs/react-api.html#reactchildrentoarray

ciampo replied:
I thought so as well. I'll close my [other comment](https://github.com/WordPress/gutenberg/pull/35532#discussion_r726916888) in favour of this one

andrewserong replied:
Thanks for the link, I hadn't used that particular API before, good to know! I've updated the function to cast as an array before concatenating strings.

gziolo commented on packages/element/README.md:
```diff
@@ -294,7 +294,7 @@ aside from `children` are passed.
 _Parameters_
 
--   _props_ `RawHTMLProps`: Children should be a string of HTML. Other props will be passed through to div wrapper.
+-   _props_ `RawHTMLProps`: Children should be a string of HTML or an array of strings. Other props will be passed through to the div wrapper.
```
It would be great to cover this change also in the CHANGELOG file of the `@wordpress/element` package:
https://github.com/WordPress/gutenberg/blob/trunk/packages/element/CHANGELOG.md

andrewserong replied:
Thanks, I've added an entry. Please let me know if it needs to be more descriptive.

andrewserong committed:
Cast children as an array, check that each child is a string before concatenating, add extra tests

andrewserong commented:
Thanks for the reviews and feedback @gziolo and @ciampo! 🙇 I've updated this PR to cast the children as an array via `Children.toArray()` and then check that each element is a string before concatenating. I _think_ this is consistent with the previous expectations of how the component should behave. Is this what you had in mind?

I've added a couple of additional unit tests, to check that it renders an empty container when passed no children, and that non-string-based children are ignored in the output, to hopefully harden the logic slightly. While I can imagine a use case for rendering out React components as plain inner html, it looks like this component was originally designed to just support being passed strings.

> Also, what's the need for such a component? We should make sure that it's fed raw HTML only from trusted sources (e.g. not plugins / users / unknown API responses...)

Good question, the component is used a fair bit within the repo, usually with trusted sources, and I see a couple of instances where the HTML is passed to a `safeHTML` function before being rendered by the `RawHTML` component, e.g. [here](https://github.com/wordpress/gutenberg/blob/69e348ec80dc2cb9d537a8ba3cf6774b48636461/packages/block-editor/src/components/block-list/block.js#L145) and [here](https://github.com/wordpress/gutenberg/blob/f0095926847286cfa32a72439b7cd8b17d253820/packages/block-editor/src/components/block-compare/block-view.js#L28).

Let me know if either of you would like to see any further changes!

gziolo commented:
This failing e2e test needs to be investigated:

<img width="963" alt="Screen Shot 2021-10-13 at 10 39 43" src="https://user-images.githubusercontent.com/699132/137098285-d15af76a-dc9f-4a5e-ae4c-0792f6310de7.png">

ntsekouras commented:
>This failing e2e test needs to be investigated:

This is not related and has been fixed in trunk. Just rebase.

gziolo commented:
> > This failing e2e test needs to be investigated:
> 
> This is not related and has been fixed in trunk. Just rebase.

I have just seen it in another PR, but you were faster with reporting it back 👍🏻

ciampo commented on packages/element/src/raw-html.js:
```diff
@@ -12,16 +12,24 @@ import { createElement } from './react';
+	// Cast children as an array, and concatenate each element if it is a string.
+	Children.toArray( children ).forEach( ( child ) => {
+		if ( typeof child === 'string' ) {
+			rawHtml += child.trim();
```
Not sure if trimming white spaces could lead to unexpected changes (see [example](https://codepen.io/ciampo/pen/porJxvX))

andrewserong replied:
Ah, good call. The logic I was after was to only add to the `rawHtml` if after trimming, the string isn't empty. I'll move the `trim` to the if statement instead.

andrewserong replied:
I've updated this now, so the logic is how I'd personally expect it to behave (no gaps unexpectedly added in the `innerHTML`, but the actual strings passed to RawHTML are not trimmed). I don't feel too strongly about keeping the `trim()` check in, though, if folks think it's better to remove it.

gziolo replied:
Yes, it's hard to tell what would be the best option. We can always leave it as is and update later when someone runs into an issue 😅

ciampo reviewed (commented):
> I've updated this PR to cast the children as an array via Children.toArray() and then check that each element is a string before concatenating. I think this is consistent with the previous expectations of how the component should behave. Is this what you had in mind?

I'll let @gziolo have the last word here, but it looks good to me

andrewserong committed:
Move trim() to the if statement to only skip empty children rather than trim passed html

andrewserong commented:
> This is not related and has been fixed in trunk. Just rebase.

Thanks, that did the trick, the e2es are passing now. I also did a tiny update to the `trim()` logic following on from Marco's feedback. Happy to remove it altogether if folks would prefer.

gziolo commented on packages/element/src/raw-html.js:
```diff
@@ -12,16 +12,24 @@ import { createElement } from './react';
  * @return {JSX.Element} Dangerously-rendering component.
  */
 export default function RawHTML( { children, ...props } ) {
-	// The DIV wrapper will be stripped by serializer, unless there are
```
Should we leave this comment? It sounds like something important.

ciampo replied:
It was removed because it didn't reflect the behaviour of the component — see https://github.com/WordPress/gutenberg/pull/35532#discussion_r726914817

gziolo replied:
It's still true:

https://github.com/WordPress/gutenberg/blob/80112e467d7ff6e537d7229e1aae559a518c2a92/packages/element/src/serialize.js#L387-L396

ciampo replied:
I see. We both got confused by the comment and thought that by "serialiser" the comment referred to React's internal serialiser when calling `createElement` a few lines below.

We should put this comment back (and maybe make it more explicit about what is meant by "serializer")

gziolo replied:
Yes, it should be further clarified.

andrewserong replied:
Excellent, thanks for clarifying. I've restored the comment and reworded it so that it's clear which serializer we're talking about. Glad we got to the bottom of that!

gziolo reviewed (approved):
@andrewserong, I think it looks good and it should remain backward compatible.

andrewserong committed:
Restore comment about the div wrapper being stripped, re-word to clarify which serializer we mean

gziolo commented:
Thank you @andrewserong and @ciampo for wrangling this enhancement 👍🏻

andrewserong commented:
Pleasure, thanks for the review and feedback! 🙇
